trajopt/src/collision_terms.cpp

```diff
@@ -421,6 +454,295 @@ void SingleTimestepCollisionEvaluator::Plot(const tesseract_visualization::Visua
+                                                       SafetyMarginData::ConstPtr safety_margin_data,
+                                                       tesseract_collision::ContactTestType contact_test_type,
+                                                       double longest_valid_segment_length,
+                                                       sco::VarVector vars0,
+                                                       sco::VarVector vars1,
+                                                       CollisionExpressionEvaluatorType type)
+  : CollisionEvaluator(std::move(manip),
+                       std::move(env),
+                       std::move(adjacency_map),
+                       world_to_base,
+                       std::move(safety_margin_data),
+                       contact_test_type,
+                       longest_valid_segment_length)
+
+  , type_(type)
+{
+  vars0_ = std::move(vars0);
+  vars1_ = std::move(vars1);
+
+  contact_manager_ = env_->getDiscreteContactManager();
+  contact_manager_->setActiveCollisionObjects(adjacency_map_->getActiveLinkNames());
+  contact_manager_->setContactDistanceThreshold(safety_margin_data_->getMaxSafetyMargin());
+
+  state_solver_ = env_->getStateSolver();
+
+  switch (type_)
+  {
+    case CollisionExpressionEvaluatorType::START_FREE_END_FREE:
+    {
+      fn_ = std::bind(
+          &DiscreteCollisionEvaluator::CalcDistExpressionsBothFree, this, std::placeholders::_1, std::placeholders::_2);
+      break;
+    }
+    case CollisionExpressionEvaluatorType::START_FIXED_END_FREE:
+    {
+      fn_ = std::bind(
+          &DiscreteCollisionEvaluator::CalcDistExpressionsEndFree, this, std::placeholders::_1, std::placeholders::_2);
+      break;
+    }
+    case CollisionExpressionEvaluatorType::START_FREE_END_FIXED:
+    {
+      fn_ = std::bind(&DiscreteCollisionEvaluator::CalcDistExpressionsStartFree,
+                      this,
+                      std::placeholders::_1,
+                      std::placeholders::_2);
+      break;
+    }
+  };
+}
+
+void DiscreteCollisionEvaluator::CalcCollisions(const DblVec& x, tesseract_collision::ContactResultVector& dist_results)
+{
+  Eigen::VectorXd s0 = sco::getVec(x, vars0_);
+  Eigen::VectorXd s1 = sco::getVec(x, vars1_);
+  tesseract_collision::ContactResultMap contact_results;
+
+  // The first step is to see if the distance between two states is larger than the longest valid segment. If larger
+  // the collision checking is broken up into multiple casted collision checks such that each check is less then
+  // the longest valid segment length.
+  double dist = (s1 - s0).norm();
+  long cnt = 2;
+  if (dist > longest_valid_segment_length_)
+  {
+    // Calculate the number state to interpolate
+    cnt = static_cast<long>(std::ceil(dist / longest_valid_segment_length_)) + 1;
+  }
+
+  // Get active link names
+  const std::vector<std::string>& active_links = adjacency_map_->getActiveLinkNames();
```

Comment:
This is not in the CastCollisionEvaluator. Should it be added around 812?

Reply:
This has been restructured.